Add AddControllers() and AddRazorPages()

rynowak opened the pull request:
These are *new style* for configuring MVC in services. We're adding these to make
things feel a bit more tailored to those particular scenarios.

----

The main reason for this is that we've had repeated community asks for an
*API-optimized* way of configuring MVC. I don't think that using AddMvcCore
is a suitable building block, because it has too many options that you want.

I've think I've identified the reasonable set of features that should be part
of the default experience for APIs. All of these things are already pay-for-play
and are activated by the presence of attributes. The only additional cost is
loading of assemblies and a few additional inspections of the attributes
(cached).

----

Given that we're creating AddControllers, it makes sense that we'd flesh out the
rest of the space that folks are saying *but what about me???*.

**Edit:** (and then there were three)

You can either:
- `AddControllers`
- `AddControllersWithViews`
- `AddRazorPages`

These are all top-level service collection methods, that return the existing `IMvcBuilder`. You can use multiple of these. If you end up using `AddControllers` and `AddRazorPages` you will end up with what `AddMvc` does today.

If you do multiple things to multiple different MVC builder instances, then they run in the order you specified - like what would happen if you use `AddMvc` multiple times in today's landscape. You can't have separate model binders for controllers and pages, or things like that.


*Small note - `AddRazorPages` includes the non-API-focused support for controllers. We'd have to do significant work to take this out. Will do it in the future based on user feedback.*

----

Branding and perception are important, and we've had users ask for more flexibility
in what shows up in their app. The plan is to update the templates to use this experience
in preview4, and see what kind of feedback we get.

rynowak committed:
Add AddControllers and AddRazorPages

These are *new style* for configuring MVC in services. We're adding these to make
things feel a bit more tailored to those particular scenarios.

----

The main reason for this is that we've had repeated community asks for an
*API-optimized* way of configuring MVC. I don't think that using AddMvcCore
is a suitable building block, because it has too many options that you want.

I've think I've identified the reasonable set of features that should be part
of the default experience for APIs. All of these things are already pay-for-play
and are activated by the presence of attributes. The only additional cost is
loading of assemblies and a few additional inspections of the attributes
(cached).

----

Additionally the AddControllers experience is composible. You can add views
to it, or add pages and get the whole thing.

AddRazorPages is basically an alias for what AddMvc does today. We don't
currently have a way to add pages without controllers (the opposite is true).
Based on feedback we could specialize this more.

----

Branding and perception are important, and we've users ask for more flexibility
in what gets added. The plan is to update the templates to use this experience
in preview4, and see what kind of feedback we get.

rynowak commented on src/Mvc/Mvc.Core/src/Builder/ControllerEndpointRouteBuilderExtensions.cs:
```diff
@@ -106,7 +106,7 @@ public static IEndpointConventionBuilder MapDefaultControllerRoute(this IEndpoin
 
         /// <summary>
         /// Adds endpoints for controller actions to the <see cref="IEndpointRouteBuilder"/> and specifies a route
-        /// with the given <paramref name="name"/>, <paramref name="areaName"/>, <paramref name="template"/>, 
```
VS is doing a format.

ryanbrandenburg replied:
For anyone who doesn't already know: adding `?w=1` to the end of the url hides whitespace only diff.

rynowak commented on src/Mvc/Mvc.ViewFeatures/src/ViewComponentResult.cs:
```diff
@@ -64,7 +64,15 @@ public override Task ExecuteResultAsync(ActionContext context)
 
             var services = context.HttpContext.RequestServices;
-            var executor = services.GetRequiredService<IActionResultExecutor<ViewComponentResult>>();
+            var executor = services.GetService<IActionResultExecutor<ViewComponentResult>>();
```
So one of the things that I added is some guardrails around things like returning `View()` from a controller without the services. I figure this is is the most likely way for someone to experience this particular problem.

I couldn't think of other things that need to be fixed, there aren't many APIs that the user calls that wouldn't work, it's mainly the view result stuff.

Additionally, Identity UI already registers what it needs in services, so we're good there.

pranavkm replied:
We could always add more if we get user reports for `GetRequiredService` failing.

rynowak commented on src/Mvc/Mvc/src/MvcServiceCollectionExtensions.cs:
```diff
@@ -30,69 +31,156 @@ public static IMvcBuilder AddMvc(this IServiceCollection services)
+                .AddAuthorization()
+                .AddCors()
+                .AddDataAnnotations()
+                .AddFormatterMappings();
```
These are all of the things that you'll get by default. 
- API explorer doesn't do anything unless you call its APIs
- Auth is based on attributes
- Cors is based on attributes
- Data annotations is based on attributes
- Format filter is based on attributes

All of them are pay for play. 

I debated leaving out format filter because no one uses it. But I think it's hard to explain why we left it out that it is to just put it in. I still find it super duper vexxxing that format filter requires a gesture.

rynowak replied:
I'm not sure. I think it's up to them to decide if this is what they wanted all along.

ryanbrandenburg reviewed (approved):
Makes sense from a templating perspective.

pranavkm commented on src/Mvc/Mvc/src/MvcServiceCollectionExtensions.cs:
```diff
@@ -30,69 +31,156 @@ public static IMvcBuilder AddMvc(this IServiceCollection services)
             return builder;
         }
 
+        internal static void AddDefaultFrameworkParts(ApplicationPartManager partManager)
```
Unrelated, but do you know what this is for?

rynowak replied:
Adding default framework parts?

pranavkm replied:
Yeah, but why do we add these? These aren't sources for controllers.

rynowak replied:
I cleaned this up by renaming this method and making sure it's only included by pages/views.

pranavkm reviewed (approved):
Do you have a follow up item to fix up the templates?

rynowak commented:
@ryanbrandenburg @pranavkm - can you please take a second look at this? I changed how this works based on feedback from @DamianEdwards  and @davidfowl and updated the PR description

pranavkm commented:
Code check failed

rynowak commented:
> Code check failed

That is probably true.

pranavkm commented on src/ProjectTemplates/Web.ProjectTemplates/content/StarterWeb-FSharp/Startup.fs:
```diff
@@ -22,7 +22,8 @@ type Startup private () =
     member this.ConfigureServices(services: IServiceCollection) =
         // Add framework services.
-        services.AddMvc().AddNewtonsoftJson().AddRazorRuntimeCompilation() |> ignore
+        services.AddControllersWithViews().AddNewtonsoftJson().AddRazorRuntimeCompilation() |> ignore
```
Wouldn't this be `AddRazorPages()`?

rynowak replied:
Starterweb for F# uses both

ryanbrandenburg reviewed (approved):
Looks good as far as I can tell.

rynowak commented:
@pranavkm any more feedback?

pranavkm commented:
:shipit:
